### PR TITLE
[4.0] Postgresql save category

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue: Save category fail on pdo postgresql

### Summary of Changes
fix sql in update 
fix unlock

### Testing Instructions
create/save a category

### CODE REVIEW

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on f9ba915f63cefd17703c8b8fece2b03b99afe5fb<br /><br />Was not easy to test because

    1. Had to apply PR #24747 to be able to have installation on Postgresql for the test.
    2. Patchtester does not work with Postgresql
    3. Had to apply changes of this PR manually because branch or this PR needs a rebase to current 4.0-dev.

But at the end test has shown that this PR here corrects error as described.

The unlock fix might even correct other issues not observed yet.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24319">issues.joomla.org/tracker/joomla-cms/24319</a>.</sub>

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on f9ba915f63cefd17703c8b8fece2b03b99afe5fb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24319">issues.joomla.org/tracker/joomla-cms/24319</a>.</sub>

**Comment by ghost** [comment]
Status "Ready To Commit".

File: libraries/src/Table/Nested.php [hunk 1]
```diff
@@ -1585,10 +1585,11 @@ protected function recursiveUpdatePublishedColumn($pk, $newState = null)
 
 		// Update and cascade the publishing state.
 		$query->clear()
-			->update("$table AS c")
-			->innerJoin("($subquery) AS c2 ON c2.newId = c.$key")
-			->set("$published = c2.newPublished")
-			->where("c.$key IN (" . implode(',', $pks) . ")");
+			->update("$table")
```
Comment (wilsonge) [inline libraries/src/Table/Nested.php]: You can remove the quotes around `$table`

File: libraries/src/Table/Nested.php [hunk 2]
```diff
@@ -1585,10 +1585,11 @@ protected function recursiveUpdatePublishedColumn($pk, $newState = null)
...
+			->innerJoin("($subquery) AS c2")
+			->set("$published = " . $this->_db->quoteName("c2.newpublished"))
+			->where("$key = c2.newId")
+			->where("$key IN (" . implode(',', $pks) . ")");
```
Comment (wilsonge) [inline libraries/src/Table/Nested.php]: can we use the `whereIn` thing here please

File: libraries/src/Table/Nested.php [hunk 3]
```diff
@@ -1585,10 +1585,11 @@ protected function recursiveUpdatePublishedColumn($pk, $newState = null)
...
 		// Update and cascade the publishing state.
 		$query->clear()
-			->update("$table AS c")
-			->innerJoin("($subquery) AS c2 ON c2.newId = c.$key")
-			->set("$published = c2.newPublished")
-			->where("c.$key IN (" . implode(',', $pks) . ")");
+			->update("$table")
+			->innerJoin("($subquery) AS c2")
+			->set("$published = " . $this->_db->quoteName("c2.newpublished"))
+			->where("$key = c2.newId")
```
Comment (wilsonge) [inline libraries/src/Table/Nested.php:1591]: I don't understand this change. What was the issue with the `ON` query in the join?

Comment (alikon) [reply libraries/src/Table/Nested.php:1591]: may help this article https://makandracards.com/makandra/32357-postgresql-vs-mysql-how-to-update-using-a-join

File: libraries/src/Table/Nested.php (hunk 1 above)
Comment (twister65) [reply libraries/src/Table/Nested.php]: `->update($this->_db->quoteName($table))`
should be more adequate.

Comment (twister65) [reply libraries/src/Table/Nested.php]: Sorry, we already have:
`$table     = $this->_db->quoteName($this->_tbl);`

File: libraries/src/Table/Nested.php (hunk 2 above)
Comment (twister65) [reply libraries/src/Table/Nested.php]: `->whereIn($key, $pks);`

File: libraries/src/Table/Nested.php (hunk 1 above)
Comment (alikon) [reply libraries/src/Table/Nested.php]: done

File: libraries/src/Table/Nested.php (hunk 2 above)
Comment (alikon) [reply libraries/src/Table/Nested.php]: done

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on b2d61aff1f8957c94c61d7fee8f797e1d0c0d2e8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24319">issues.joomla.org/tracker/joomla-cms/24319</a>.</sub>

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on b2d61aff1f8957c94c61d7fee8f797e1d0c0d2e8<br /><br />with MySQL and PostgreSQL.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24319">issues.joomla.org/tracker/joomla-cms/24319</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**